### ISSUE TITLE
FR: Fast user switching

### ISSUE THREAD

**Issue by apenwarr** [tailscale/tailscale#713]
(context added on 2022/09/14)

### What are you trying to do?
Log into multiple tailscale accounts at once (eg. example@gmail.com and example@example.com) and easy switch between accounts, i.e. improve the "Log in to a different account..." flow on a device, (possibly) without requiring re-authentication. Access only machines on the network you are currently logged into.

**FR to access all the machines on both networks simultaneously: https://github.com/tailscale/tailscale/issues/183**

### How should we solve this?

The basics of this are not that complicated to do purely client side:
- have ipn.Backend start multiple controlclient instances, one per login
- merge the netmaps from the two instances and feed the merged result to wgengine

Complexity arises from specific details:

1) Should we have multiple local IP addresses, one per account? Is that even possible on all platforms? Or should we get help from the control server to assign the same IP address to all accounts? But if we do that, what about accounts with custom IP subnet designations?

2) If we have multiple local IP addresses, there needs to be a good way to choose which local IP address to use for a given outgoing connection. (Based on routes?)

3) If we use a single local IP address, this could violate users' privacy expectations. Their distinctive identities end up tied together in the control server. It's a lot more elegant if we can support multi-login entirely client side.

4) If two accounts export the same subnet routes (--advertise-subnets), you can get conflicts between them.

5) If two accounts have different DNS overrides, they can conflict. (This isn't so bad if we use the DNS settings only for particular subdomains rather than globally.)

6) Corporate admins may not like the idea of a single machine being on both the corporate and non-corporate networks at once (although it's probably not as bad as being on both the Internet and the corporate network at once.)

**Comment by happyalu** [comment]
Although not _at the same time_, I find it useful to be able to switch accounts to access different groups of machines from my laptop. Prior to 1.0, with the [tailscale-linux-switchuser](https://github.com/happyalu/tailscale-linux-switchuser/) tool I was saving copies of relay.conf for different email addresses and setting `relay.conf` to the required account. It seems that now the information is saved in `tailscaled.state` with a base64 encoded field. 

I tried to save this new file for different email addresses similarly, and copy the right one for the address I want to login to tailscale with, but that didn't seem to work even after restarting tailscaled. Is there an easy way to switch accounts in the current setup? I thought I'll ask before trying to dive down into finding the details of the new file :)

**Comment by apenwarr** [comment]
Note: related issue: https://github.com/tailscale/tailscale/issues/183

**Comment by apenwarr** [comment]
@happyalu, a workaround for user switching is to install the latest *unstable* tailscale stream (at least version 1.1.372) and try the `tailscale up --force-reauth` option. This should let you switch between accounts more easily than by switching config files entirely.

**Comment by DentonGentry** [comment]
Related issue: https://github.com/tailscale/tailscale/issues/662

**Comment by noopsy** [comment]
`tailscale up --force-reauth` will create a new machine and with a new tailscale IP in the tailscale admin panel.  I have found that I need to delete the old machine in the tailscale admin panel before I can get network connectivity to/from this machine.

**Comment by pirate** [comment]
You can already sort of achieve this by running two separate tailscale instances in docker-compose, but YMMV as I haven't thoroughly tested this in production (only the variant described below):

**`docker-compose.yml`:**
```yaml
version: '2.4'
services:
    tailscale1:
        image: tailscale:latest
        build: https://github.com/tailscale/tailscale.git#v1.4.4
        command: tailscaled
        cap_add:
            - NET_ADMIN
            # - SYS_MODULE             # usually not needed, depends on host OS
        network_mode: host
        volumes:
            - /dev/net/tun:/dev/net/tun
            - ./data/tailscale1:/var/lib/tailscale
            
     tailscale2:
        image: tailscale:latest
        command: tailscaled
        cap_add:
            - NET_ADMIN
            # - SYS_MODULE             # usually not needed, depends on host OS
        network_mode: host
        volumes:
            - /dev/net/tun:/dev/net/tun
            - ./data/tailscale2:/var/lib/tailscale
```
```bash
# then start the containers and log into your two separate tailscale accounts
docker-compose up
docker-compose exec tailscale1 tailscale up
docker-compose exec tailscale2 tailscale up
```

Alternatively, you can remove the `network_mode: host` line to keep the containers in a separate network (shared between the containers, but isolated from the host). In that setup each machine will appear to be connected to their respective separate tailscale network but also connected to a shared LAN with the other container.

**Comment by hkrutzer** [comment]
I see https://github.com/tailscale/tailscale/issues/1129 was closed as a duplicate of this, which also mentions fast user switching. Is that also covered by this issue? For me connecting to multiple networks at the same time is not that important, but switching between my personal and corporate account is very annoying.

**Comment by bradfitz** [comment]
@hkrutzer, yes, it's the same client work required for both.

**Comment by hkrutzer** [comment]
OK, I thought this issue might be different because I don't think you would encounter "Should we have multiple local IP addresses, one per account?" or any of the other bullet points for faster user switching 🙂

**Comment by adrm** [comment]
Being able to easily switch accounts would be enough for most cases, much better than the situation we have right now. I am commenting because this issue seems to contemplate a much more complicated use-case that might not be as needed, and can come after the ability to switch accounts. This would be life-changing for those of us using Tailscale for personal use and for corporate.

**Comment by nitzanm** [comment]
I'll agree with @adrm on that - just the ability to log in to multiple accounts and switch between them with a click would be a good enough solution that it probably wouldn't be worth the additional 90% of the work to make accessing multiple networks at once work (especially given the complexities in the bullet points in the issue).

My hacky workaround for now is to [share my personal machines](https://tailscale.com/kb/1084/sharing/) with my corporate Tailscale user, and then log in with the corporate user on the Tailscale client. This does have privacy implications - for example, corporate IT could change my password, log in as me, and access my personal machines. But I find it unlikely enough that I'm accepting the risk. Plus, all my personal machines need additional authentication (SSH key etc.), and I find it unlikely that corporate IT is going to start brute forcing my keys/passwords.

If anyone sees a hole in that workaround that I'm missing, I'd love to hear it!

**Comment by alie2n** [comment]
+1 for being able to switch between accounts.
My use case would be to join my mobile phone (ios) to multiple Tailscale networks (work, customer1, private) and switch between these as necessary.

**Comment by bilby91** [comment]
Is this a priority at the moment? Looking forward on this feature!

**Comment by bradfitz** [comment]
The existence of this bug implies that it's a priority. There are many bugs and many priorities. This is not the top priority. But we're making progress towards this one by working on other necessary stuff first.

**Comment by bilby91** [comment]
@bradfitz Thanks for the update!

**Comment by sanderversluys** [comment]
Really looking forward to this feature. Simple switching between accounts, not simultaneous use, would be a perfect solution for many i guess. At the moment I have a corporate account and personal business accounts.

**Comment by filviu** [comment]
I think I'm against the grain. I see most people testing Tailscale at home and then using it at work. I went the other way around. I'm implementing it at work and holy $#!^ it's fast and easy to use. So I want it now at home too. A simple drop-down to switch between accounts would be awesome.

**Comment by bithavoc** [comment]
@bradfitz I think Tailscale is throwing away part of the acquisition funnel here, if a developer o devops starts using tailscale at personal level and re-uses the equipment between clients and personal setups, they won't recommend Tailscale it to clients because that would ruin their personal setup given there is no way to use multiple network or switch easily between accounts.

**Comment by Gurkengewuerz** [comment]
+1 on this bust with a quick reminder:
[juanfont/headscale](https://github.com/juanfont/headscale) by @juanfont is getting bigger and bigger currently with over 7.3k stars. We have a workaround to configure the endpoint just like in the newest Android client by using the registry on Windows.
So **please** keep in mind users exists that are running their own instances and need to configure *multiple* endpoints. 🙌🏼

**Comment by bithavoc** [comment]
@Gurkengewuerz does that mean I can deploy multiple agents on the same server pointing to multiple tailscale/head scale accounts ACL’s?

**Comment by Gurkengewuerz** [comment]
> @Gurkengewuerz does that mean I can deploy multiple agents on the same server pointing to multiple tailscale/head scale accounts ACL’s?

Well what i mean is just the ability to switch between accounts (like personal and business) with the ability to set endpoints per account i.e. the personal account points to `controlplane.gurkengewuerz.me` and the business account points to `controlplane.my-business.com`.

Multiple agents on the same server as you described sounds to me like another issue #183

**Comment by jace** [comment]
> I think Tailscale is throwing away part of the acquisition funnel here, if a developer o devops starts using tailscale at personal level and re-uses the equipment between clients and personal setups, they won't recommend Tailscale it to clients because that would ruin their personal setup given there is no way to use multiple network or switch easily between accounts.
> 

That is exactly my dilemma today. I'm about to head out on my first post-pandemic trip. My ISP has CGNA (but 10.x IP range), so going outdoors means I will lose complete access to my now-elaborate WFH setup.

Tailscale has solved this problem and made me a very happy fellow, and I want to introduce it at work, except that means I have to introduce all my colleagues into my home network or make a separate work network and deal with the pain of switching between them.

**Comment by DentonGentry** [comment]
Implementation of this feature has begun in earnest.

**Comment by mihaip** [comment]
Fast user switching is now available in Linux, Mac and Windows [unstable builds](https://tailscale.com/kb/1083/install-unstable/). Feedback (as comments on this issue) or bug reports (as separate issues) is appreciated.

In the UI apps, it is available as "Add another account…" in the status menu.
| Windows | Mac |
|-|-|
| ![image](https://user-images.githubusercontent.com/513813/203432939-676d3f57-da58-4c59-b1d2-a256bcb25c1e.png) | <img width="599" alt="image (1)" src="https://user-images.githubusercontent.com/513813/203432957-5a24dfc6-e718-4fbb-9149-cca1e290d10b.png"> |
 
When using the `tailscale` command line app, you can create a new profile via `tailscale login` and switch between them with `tailscale switch` (`tailscale switch --list` shows all profiles).

**Comment by Gurkengewuerz** [comment]
Any plans to support setting **multiple** `LoginURL`s? 👀
In Windows Tailscale uses `Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Tailscale IPN\LoginURL` for the main account. If i add another account it gets ignored.

**Comment by maisem** [comment]
You can use `tailscale login --login-server=...` to use a custom coordination server.

**Comment by juanfont** [comment]
@maisem I reckon they mean to be connected _at the same time_ to multiple coordination servers ;)

**Comment by mihaip** [comment]
Fast user switching has now shipped in Tailscale 1.34. It is available for the CLI on Linux, Mac and Windows, and in the Mac and Windows UI clients, see the blog post for a summary: https://tailscale.com/blog/fast-user-switching/. 

There is also more complete documentation for the feature at https://tailscale.com/kb/1225/fast-user-switching/.

Thanks for your patience for this long-awaited feature. If there are any bugs or additional platforms or feedback to consider, please file them as separate GitHub issues (instead of extending this one).
